File: src/Tests/Behavioral/GenericCompositeLiterals/GenericCompositeLiterals.go
package main

import "fmt"

// Queue is a generic queue implementation
type Queue[T any] struct {
	items []T
}

func (q *Queue[T]) Enqueue(item T) {
	q.items = append(q.items, item)
}

func (q *Queue[T]) Dequeue() (T, bool) {
	var zero T
	if len(q.items) == 0 {
		return zero, false
	}

	item := q.items[0]
	q.items = q.items[1:]
	return item, true
}

func (q *Queue[T]) Size() int {
	return len(q.items)
}

func main() {
	// Empty composite literal with generic type (CompositeLit)
	intQueue := Queue[int]{}
	intQueue.Enqueue(10)
	intQueue.Enqueue(20)
	fmt.Printf("Int queue size: %d\n", intQueue.Size())

	// Composite literal with field values
	stringQueue := Queue[string]{
		items: []string{"hello", "world"},
	}
	fmt.Printf("String queue size: %d\n", stringQueue.Size())

	// Composite literal for a generic type within a variable declaration
	var floatQueue = Queue[float64]{
		items: []float64{3.14, 2.71, 1.618},
	}
	fmt.Printf("Float queue size: %d\n", floatQueue.Size())

	// Multiple instantiations in a slice literal
	queues := []interface{}{
		Queue[int]{},
		Queue[string]{},
		Queue[bool]{items: []bool{true, false}},
	}
	fmt.Printf("Number of queues: %d\n", len(queues))
}

File: src/Tests/Behavioral/GenericCompositeLiterals/GenericCompositeLiterals.cs
namespace go;

using fmt = fmt_package;

partial class main_package {

[GoType] partial struct Queue<T> {
    internal slice<T> items;
}

[GoRecv] public static void Enqueue<T>(this ref Queue<T> q, T item) {
    q.items = append(q.items, item);
}

[GoRecv] public static (T, bool) Dequeue<T>(this ref Queue<T> q) {
    T zero = default!;
    if (len(q.items) == 0) {
        return (zero, false);
    }
    var item = q.items[0];
    q.items = q.items[1..];
    return (item, true);
}

[GoRecv] public static nint Size<T>(this ref Queue<T> q) {
    return len(q.items);
}

internal static void Main() {
    var intQueue = new Queue<nint>(nil);
    intQueue.Enqueue(10);
    intQueue.Enqueue(20);
    fmt.Printf("Int queue size: %d\n"u8, intQueue.Size());
    var stringQueue = new Queue<@string>(
        items: new @string[]{"hello", "world"}.slice()
    );
    fmt.Printf("String queue size: %d\n"u8, stringQueue.Size());
    Queue<float64> floatQueue = new Queue<float64>(
        items: new float64[]{3.14D, 2.71D, 1.618D}.slice()
    );
    fmt.Printf("Float queue size: %d\n"u8, floatQueue.Size());
    var queues = new any[]{
        new Queue<nint>(nil),
        new Queue<@string>(nil),
        new Queue<bool>(items: new bool[]{true, false}.slice())
    }.slice();
    fmt.Printf("Number of queues: %d\n"u8, len(queues));
}

} // end main_package
